### PR TITLE
Scatter update implemented with in-place memory, gradient_check modified a little bit to suit in-place

### PR DESCRIPTION
**Pull request by zchen0211**
I implemented this function.
https://www.tensorflow.org/api_docs/python/tf/scatter_update
with the add case. Use in-place memory for forward and backward, with Python unit-test passed.
Also, I modified the gradient check a little bit (reset data back to origin) to make sure the numerical gradient is computed correctly.

### CODE REVIEW

File: python/paddle/v2/framework/tests/gradient_checker.py [hunk 1]
```diff
@@ -90,6 +90,9 @@ def product(dim):
     # we only compute gradient of one element each time.
     # we use a for loop to compute the gradient of every element.
     for i in xrange(tensor_size):
+        for var_name in input_values:
```
Comment (jacquesqiao) [inline python/paddle/v2/framework/tests/gradient_checker.py]: add a parameter to choose if use inplace or not to speed up gradient check

Comment (zchen0211) [reply python/paddle/v2/framework/tests/gradient_checker.py]: Thx. Already added, with in-place option = False as default.

File: paddle/operators/scatter_op.cc [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+namespace operators {
+
+class ScatterOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    framework::DDim output_dims(ctx.Input<Tensor>("Ref")->dims());
+    ctx.Output<Tensor>("Out")->Resize(output_dims);
```
Comment (jacquesqiao) [inline paddle/operators/scatter_op.cc]: 感觉这里也可以写成一行
```cpp
ctx.Output<Tensor>("Out")->Resize(ctx.Input<Tensor>("Ref")->dims());
```

File: paddle/operators/scatter_op.cc [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  }
+};
+
+class ScatterGradOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto Updates_grad = ctx.Output<Tensor>(framework::GradVarName("Updates"));
```
Comment (jacquesqiao) [inline paddle/operators/scatter_op.cc]: we suggest `auto *` for pointer type

File: paddle/operators/scatter_op.cu [hunk 4]
```diff
@@ -0,0 +1,20 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
```
Comment (jacquesqiao) [inline paddle/operators/scatter_op.cu]: the format of this copyright is different with others.

File: paddle/operators/scatter_op.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    // Apply ScatterUpdate: Out[index] += Updates[:]
+    ScatterUpdate<T>(ctx.GetPlace(), Updates, Index, Out);
+  }
+};
+
+template <typename Place, typename T>
+class ScatterGradientOpKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext &ctx) const override {
+    auto *dRef = ctx.Output<Tensor>(framework::GradVarName("Ref"));
```
Comment (jacquesqiao) [inline paddle/operators/scatter_op.h]: in scatter_op.cc the name is xx_grad, here is dxx, may be we should use a coincident style

File: paddle/operators/scatter_op.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    auto *dUpdates = ctx.Output<Tensor>(framework::GradVarName("Updates"));
+    auto *Index = ctx.Input<Tensor>("Index");
+    auto *dO = ctx.Input<Tensor>(framework::GradVarName("Out"));
```
Comment (jacquesqiao) [inline paddle/operators/scatter_op.h]: dO is strange, looks like d0(zero)

File: paddle/operators/scatter_op.cc [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    auto Updates = ctx.Input<Tensor>("Updates");
+    auto Ref_grad = ctx.Output<Tensor>(framework::GradVarName("Ref"));
+    auto Ref = ctx.Input<Tensor>("Ref");
+
+    Ref_grad->Resize(Ref->dims());
```
Comment (jacquesqiao) [inline paddle/operators/scatter_op.cc]: maybe should check the dim of the inputs.

File: python/paddle/v2/framework/tests/gradient_checker.py [hunk 8]
```diff
@@ -99,6 +104,10 @@ def product(dim):
         y_pos = get_output()
 
         # plus delta to this element, run op and get the sum of the result tensor.
+        if in_place:
```
Comment (jacquesqiao) [inline python/paddle/v2/framework/tests/gradient_checker.py]: may be should move this code to a function
```python
def restore_inputs(input_values, scope):
        for var_name in input_values:
             tensor_ = scope.find_var(var_name).get_tensor()
             tensor_.set(numpy.copy(input_values[var_name]), core.CPUPlace())
```

File: python/paddle/v2/framework/tests/test_scatter_op.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    def test_scatter_grad(self):
+        op = create_op("scatter")
+        # test data setup
+        ref_np = numpy.ones((3, 10)).astype("float32")
+        index_np = numpy.array([1, 2]).astype("int32")
+        updates_np = numpy.random.random((2, 10)).astype("float32")
+        output_np = numpy.copy(ref_np)
+        output_np[index_np] += updates_np
+        inputs = {'Ref': ref_np, 'Index': index_np, 'Updates': updates_np}
+        # check gradient
```
Comment (jacquesqiao) [inline python/paddle/v2/framework/tests/test_scatter_op.py]: remove this comment

File: paddle/operators/scatter_op.cc (hunk 2 above)
Comment (jacquesqiao) [reply paddle/operators/scatter_op.cc]: 或者
```cpp
auto* Ref = ctx.Input<Tensor>("Ref");
ctx.Output<Tensor>("Out")->Resize(Ref->dims());
```

Comment (zchen0211) [reply paddle/operators/scatter_op.cc]: Done.

File: paddle/operators/scatter_op.cc (hunk 3 above)
Comment (zchen0211) [reply paddle/operators/scatter_op.cc]: done

File: paddle/operators/scatter_op.h (hunk 5 above)
Comment (zchen0211) [reply paddle/operators/scatter_op.h]: done

File: paddle/operators/scatter_op.h (hunk 6 above)
Comment (zchen0211) [reply paddle/operators/scatter_op.h]: changed to dOut already.

File: python/paddle/v2/framework/tests/test_scatter_op.py (hunk 9 above)
Comment (zchen0211) [reply python/paddle/v2/framework/tests/test_scatter_op.py]: done

File: python/paddle/v2/framework/tests/gradient_checker.py (hunk 8 above)
Comment (zchen0211) [reply python/paddle/v2/framework/tests/gradient_checker.py]: moved into copy_tensor()

File: paddle/operators/scatter_op.cu (hunk 4 above)
Comment (zchen0211) [reply paddle/operators/scatter_op.cu]: done

File: python/paddle/v2/framework/tests/gradient_checker.py (hunk 8 above)
Comment (jacquesqiao) [reply python/paddle/v2/framework/tests/gradient_checker.py]: I think copy_tensor is not a very good name. Can not know what it means by reading the name. maybe `restore_inputs` or `reset_inputs`

File: paddle/operators/scatter_op.cu (hunk 4 above)
Comment (jacquesqiao) [reply paddle/operators/scatter_op.cu]: seems not changed

File: paddle/operators/scatter_op.cc (hunk 7 above)
Comment (jacquesqiao) [reply paddle/operators/scatter_op.cc]: please use `auto *` and do `dim check`

File: paddle/operators/scatter_op.cu (hunk 4 above)
Comment (zchen0211) [reply paddle/operators/scatter_op.cu]: changed back.

**Pull request merged**